### PR TITLE
Storages: Change the element type of `use_pack` from `bool` to `RSResult` in `DMFilePackFilter`

### LINKED ISSUE #9103: Using the filtering results of rough indexes to reduce the calculation of filters

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #9103

### What is changed and how it works?

```commit-message
1. Change `std::vector<UInt8> use_packs;` to `std::vector<RSResult> pack_res;`. 
2. Add statistical data of filtering results in `ScanContext`.
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.cpp [hunk 1]
```diff
@@ -47,33 +49,21 @@ void DMFilePackFilter::init()
...
+    pack_res = handle_res;
+    auto after_pk = countUsePack();
 
     /// Check packs by read_packs
     if (read_packs)
     {
         for (size_t i = 0; i < pack_count; ++i)
         {
-            use_packs[i] = (static_cast<bool>(use_packs[i])) && read_packs->contains(i);
+            pack_res[i] = pack_res[i] && (read_packs->contains(i) ? RSResult::Some : RSResult::None);
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.cpp:61]:
```suggestion
            pack_res[i] = read_packs->contains(i) ? pack_res[i] : RSResult::None;
```

File: dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.cpp [hunk 2]
```diff
@@ -108,14 +98,47 @@ void DMFilePackFilter::init()
...
+            ++none_count;
+            break;
+        case RSResult::Some:
+            ++some_count;
+            break;
+        case RSResult::All:
+            ++all_count;
+            break;
+        default:
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "{} is invalid", magic_enum::enum_name(res));
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.cpp:133]:
```suggestion
            throw Exception(ErrorCodes::LOGICAL_ERROR, "{} is invalid", res);
```

magic_enum::enum_name always return `` here.

File: dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.h [hunk 3]
```diff
@@ -68,9 +68,10 @@ class DMFilePackFilter
         return pack_filter;
     }
 
-    inline const std::vector<RSResult> & getHandleRes() const { return handle_res; }
-    inline const std::vector<UInt8> & getUsePacksConst() const { return use_packs; }
-    inline std::vector<UInt8> & getUsePacks() { return use_packs; }
+    const RSResults & getHandleRes() const { return handle_res; }
+    const RSResults & getPackResConst() const { return pack_res; }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.h:72]: maybe we can combine `handle_res` and `pack_res`?

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/Index/RSResult.h [hunk 4]
```diff
@@ -75,6 +75,10 @@ inline RSResult operator&&(RSResult v0, RSResult v1)
     return Some;
 }
 
+inline bool isUse(RSResult res) noexcept
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/RSResult.h:78]:
```suggestion
ALWAYS_INLINE inline bool isUse(RSResult res) noexcept
```

File: dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.h (hunk 3 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/File/DMFilePackFilter.h:72]: > `pack_res` is the filter results of `rowkey_ranges && filter && read_packs`.

Add this comment to line 171

**Review by JaySon-Huang — COMMENTED** [review]
Rest LGTM

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**